src/sound/DalSoRiR2.cc

```diff
@@ -0,0 +1,374 @@
+				ymf278BusyTime = time + WAVE_REG_SELECT_DELAY;
+				opl4latch = value;
+				break;
+			case 1:
+				if ((0x08 <= opl4latch) && (opl4latch <= 0x1F)) {
+					ymf278LoadTime = time + LOAD_DELAY;
+				}
+				if ((3 <= opl4latch) && (opl4latch <= 6)) {
+					// Note: this time is so small that on
+					// MSX you never see BUSY=1 for these
+					// registers. Confirmed on real HW that
+					// also registers 3-5 are faster.
+					ymf278BusyTime = time + MEM_WRITE_DELAY;
+				} else {
+					// For the other registers it is
+					// possible to see BUSY=1, but only
+					// very briefly and only on R800.
+					ymf278BusyTime = time + WAVE_REG_WRITE_DELAY;
+				}
+				if (opl4latch == 0xf8) {
+					ymf262.setMixLevel(value, time);
+				} else if (opl4latch == 0xf9) {
+					ymf278.setMixLevel(value, time);
+				}
+				ymf278.writeReg(opl4latch, value, time);
+				break;
+			default:
+				UNREACHABLE;
+			}
+		} else {
+			// Verified on real YMF278:
+			// Writes are ignored when NEW2=0 (both register select
+			// and register write).
+		}
+	} else {
+		// FM part  0xC0-0xC7
+		switch (port & 0x03) {
+		case 0: // select register bank 0
+			opl3latch = value;
+			ymf278BusyTime = time + FM_REG_SELECT_DELAY;
+			break;
+		case 2: // select register bank 1
+			opl3latch = value | 0x100;
+			ymf278BusyTime = time + FM_REG_SELECT_DELAY;
+			break;
+		case 1:
+		case 3: // write fm register
+			ymf278BusyTime = time + FM_REG_WRITE_DELAY;
+			ymf262.writeReg(opl3latch, value, time);
+			break;
+		default:
+			UNREACHABLE;
+		}
+	}
+}
+
+byte DalSoRiR2::readMem(word addr, EmuTime::param /*time*/)
+{
+	if ((0x3000 <= addr) && (addr < 0x4000)) {
+		// SRAM frame 0
+		return sram[addr - 0x3000];
+	} else if ((0x7000 <= addr) && (addr < 0x8000) && ((regFrame[0] & DISABLE) == 0)) {
+		// SRAM frame 1
+		return sram[(addr - 0x6000) + (regFrame[0] & REG_FRAME_SEGMENT_MASK) * 0x1000];
+	} else if ((0xB000 <= addr) && (addr < 0xC000) && ((regFrame[1] & DISABLE) == 0)) {
+		// SRAM frame 2
+		return sram[(addr - 0x9000) + (regFrame[1] & REG_FRAME_SEGMENT_MASK) * 0x1000];
+	} else {
+		// read selected segment of flash
+		auto flashBank = addr >> 13;
+		auto flashSegmentAddr = addr & 0x3FFF;
+		return flash.read(flashSegmentAddr + regBank[flashBank] * 0x4000);
+	}
+	UNREACHABLE;
+}
+
+void DalSoRiR2::writeMem(word addr, byte value, EmuTime::param /*time*/)
+{
+	if ((0x3000 <= addr) && (addr < 0x4000)) {
+		// SRAM frame 0
+		sram[addr - 0x3000] = value;
+	} else if ((0x7000 <= addr) && (addr < 0x8000) && ((regFrame[0] & DISABLE) == 0)) {
+		// SRAM frame 1
+		sram[addr - 0x6000] = value;
+	} else if ((0xB000 <= addr) && (addr < 0xC000) && ((regFrame[1] & DISABLE) == 0)) {
+		// SRAM frame 2
+		sram[addr - 0x9000] = value;
+	} else if ((0x6000 <= addr) && (addr < 0x6800)) {
+		if (addr >= 0x6700) {
+			setRegCfg(value);
+		} else if (addr >= 0x6500) {
+			regFrame[(addr >> 9) & 1] = value;
+		} else if (addr < 0x6400) {
+			regBank [(addr >> 8) & 3] = value;
+		}
```

Comment:
I think this is correct, but it took me a little while to verify that it's indeed correct. E.g. it isn't obvious that [0x6400, 0x6500) has no function.
Maybe it's clearer with this structure:
```
} else if ((0x6000 <= addr) && (addr < 0x6400)) {
    regBank[..] = ..;
} else if ((0x6500 <= addr) && (addr < 0x6700)) {
    regBank[..] = ..;
} else if ((0x6700 <= addr) && (addr < 0x6800)) {
 setRegCfg(..);
} ..
```